src/LowerTriangularMatrices/lower_triangular_matrix.jl

```diff
@@ -122,36 +171,64 @@ function LowerTriangularMatrix(M::AbstractMatrix{T}) where T
     return L
 end
 
-function Base.Matrix(L::LowerTriangularMatrix{T}) where T
-    m, n = size(L)
-    M = zeros(T, m, n)
+function LowerTriangularArray(M::Array{T,3}) where T # CPU version 
+    m, n = size(M)[1:2]
+    L = LowerTriangularArray{T,3,Array{T,2}}(undef, size(M)...)
+    
     k = 0
-    @inbounds for j in 1:n
+    @inbounds for j in 1:n      # only loop over lower triangle
         for i in j:m
-            k += 1
-            M[i, j] = L[k]
+            k += 1              # next element in lower triangle
+            L[k, :] = M[i, j, :]       # and copy data into vector 
         end
     end
-    return M
+    return L
+end
+
+# helper function for conversion etc on GPU, returns indices of the lower triangle
+lowertriangle_indices(M::AbstractMatrix) = tril!(trues(size(M)))
+lowertriangle_indices(m::Integer, n::Integer) = tril!(trues((m,n)))
+
+function lowertriangle_indices(M::AbstractArray{T,N}) where {T,N}
+    @boundscheck N >= 3 || throw(BoundsError)
+
+    indices = lowertriangle_indices(getindex(M,:,:,[1 for i=1:(N-2)]...)) # TODO: this assumes 1-indexing and won't work for some fancy indexed arrays
+    indices = reshape(indices, size(indices,1), size(indices,2), [1 for i=1:(N-2)]...)
+    
+    repeat(indices, 1, 1, size(M)[3:end]...)
+end  
+
+# GPU version and fallback for higher dimensions
+function LowerTriangularArray(M::ArrayType) where {T, N, ArrayType<:AbstractArray{T,N}} 
+    m, n = size(M)[1:2]
+    LowerTriangularArray(M[lowertriangle_indices(m,n),[Colon() for i=1:N-2]...], m, n)
+end
+
+LowerTriangularArray(L::LowerTriangularArray) = LowerTriangularArray(L.data, L.m, L.n)
+LowerTriangularMatrix(M::AbstractMatrix) = LowerTriangularArray(M)
+
+function Base.Matrix(L::LowerTriangularMatrix{T}) where T
+    m, n = size(L)
+    M = zeros(T, m, n)
+    copyto!(M, L)
 end
             
-Base.copy(L::LowerTriangularMatrix{T}) where T = LowerTriangularMatrix(L)
+Base.copy(L::LowerTriangularArray{T}) where T = deepcopy(LowerTriangularArray(L)) # TODO: I enforce a deepcopy here, as the old tests want it to be a deep copy
```

Comment:
`copy` is by default flat though, right?

Reply:
Yes, Julia standard is a flat copy. With `LowerTriangularArray` you definitely need a deep copy if you want to have two independent arrays, otherwise they point to the same memory (I tried it).

Reply:
Let's define both copy and deepcopy then? Wouldn't they be defined by default for AbstractArray?

Reply:
Yep, it's actually not needed to define it explicitly. I only did it because the old tests want `copy` to be a deep copy. Is this used somewhere in the model?

Reply:
I don't think so, not sure why I/we haven't used `deepcopy` instead. If `copy` is used then presumably only in tests or in the initial conditions. Otherwise we don't want to copy our arrays ;)

Reply:
Then I remove that, and also the test that `copy` yields a deep copy.

Reply:
(unresolve because copy = deepcopy still)

Reply:
Forgot to actually remove it, sorry